File: VoidLink/ViewControllers/HostCollectionViewController.m
//
//  HostCollectionViewController.m
//  VoidLink
//
//  Created by True砖家 on 2025/5/28.
//  Copyright 2025 True砖家 @ Bilibili. All rights reserved.
//

#import "HostCollectionViewController.h"
#import "HostCardView.h"
#import "TemporaryHost.h"
#import "VoidLink-Swift.h"

static const CGFloat cellOffsetY = 20;

@implementation HostCell {
    UIViewController* parentVC;
}

- (void)prepareForReuse {
    [super prepareForReuse];
    [self.cardView removeFromSuperview];
    self.cardView = nil;
}

- (CGFloat)getHostCardSizeFactor{
    TemporaryHost* dummyHost = [[TemporaryHost alloc] init];
    HostCardView* dummyCard = [[HostCardView alloc] initWithHost:dummyHost];
    return self.contentView.bounds.size.height/dummyCard.size.height;
}

- (UIViewController *)viewController {
    UIResponder *responder = self;
    while (responder) {
        if ([responder isKindOfClass:[UIViewController class]]) {
            return (UIViewController *)responder;
        }
        responder = responder.nextResponder;
    }
    return nil;
}

- (void)assignDelegateForHostCard{
    parentVC = [self viewController].parentViewController;
    if(parentVC != nil){
        if([parentVC conformsToProtocol:@protocol(HostCardActionDelegate)]) self.cardView.delegate = (id<HostCardActionDelegate>) parentVC;
    }
}

- (void)didMoveToWindow {
    [super didMoveToWindow];
    [self assignDelegateForHostCard];
}

- (void)configureWithHost:(TemporaryHost *)host {
    if (!self.cardView) {
        self.cardView = [[HostCardView alloc] initWithHost:host andSizeFactor:[self getHostCardSizeFactor]];
        [self assignDelegateForHostCard];
        if(!self.cardView.superview){
            [self.contentView addSubview:self.cardView];
            [NSLayoutConstraint activateConstraints:@[
                [self.cardView.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor],
                [self.cardView.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:0],
            ]];
        }
    }
}

@end

@interface HostCollectionViewController () <UICollectionViewDelegateFlowLayout>
@property (nonatomic, strong, readwrite) NSMutableArray<TemporaryHost *> *items;
@property (nonatomic, strong) NSLayoutConstraint *collectionViewHeightConstraint;
@property (nonatomic, strong) NSLayoutConstraint *superViewBottomConstraint;
@end

@implementation HostCollectionViewController{
    UICollectionViewFlowLayout *layout;
    CGFloat _horizontalPadding;
}

- (instancetype)init {
    layout = [[UICollectionViewFlowLayout alloc] init];
    switch ([UIDevice currentDevice].userInterfaceIdiom) {
        case UIUserInterfaceIdiomPhone:
            _horizontalPadding = 5;
            break;
        case UIUserInterfaceIdiomPad:
        default:
            _horizontalPadding = 75;
            break;
    }
    layout.sectionInset = UIEdgeInsetsMake(7, _horizontalPadding, 0, _horizontalPadding); // 上、左、下、右的间距
    if (self = [super initWithCollectionViewLayout:layout]) {
        _interItemMinimumSpacing = 10;
        _minimumLineSpacing = 10;
        // _cellSize = CGSizeMake(100, 100);
        _items = [NSMutableArray array];

        _collectionViewHeightConstraint = [self.collectionView.heightAnchor constraintEqualToConstant:50];
        _collectionViewHeightConstraint.active = YES;
    }
    return self;
}

- (void)updateTheme {
    self.collectionView.backgroundColor = ThemeManager.hostViewBackgroundColor;
    for (HostCell *cell in [self.collectionView visibleCells]) {
        [cell.cardView updateTheme];
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.collectionView registerClass:[HostCell class] forCellWithReuseIdentifier:@"HostCell"];
    self.collectionView.alwaysBounceVertical = NO;
    self.collectionView.showsVerticalScrollIndicator = NO;
    [self updateTheme];
}


#pragma mark - Data control

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    NSLog(@"View did appear");
}

- (void)didMoveToParentViewController:(UIViewController *)parent {
    [super didMoveToParentViewController:parent];
}

- (void)addHost:(TemporaryHost *)host {
    if(![self.items containsObject:host]){
        [self.items addObject:host];
        [self.collectionView reloadData];
    }
}

- (void)removeHost:(TemporaryHost *)host {
    if([self.items containsObject:host]){
        [self.items removeObject:host];
        [self.collectionView reloadData];
    }
}

- (void)removeLastItem {
    if (self.items.count > 0) {
        [self.items removeLastObject];
        [self.collectionView reloadData];
    }
}

- (NSInteger)numberOfRowsInCollectionView{
    NSInteger itemCount = [self.collectionView numberOfItemsInSection:0];
    
    if (itemCount == 0) return 0;
    
    NSMutableSet<NSNumber *> *rowYs = [NSMutableSet set];
    for (NSInteger i = 0; i < itemCount; i++) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForItem:i inSection:0];
        UICollectionViewLayoutAttributes *attr = [layout layoutAttributesForItemAtIndexPath:indexPath];
        if (attr) {
            CGFloat y = CGRectGetMinY(attr.frame);
            [rowYs addObject:@(round(y))]; // round 防止浮点误差
        }
    }
    
    return rowYs.count;
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];

    CGFloat contentHeight = self.collectionView.collectionViewLayout.collectionViewContentSize.height;
    bool contentExceedsView = contentHeight > self.view.superview.bounds.size.height - self.view.frame.origin.y;
    if(contentExceedsView){
        if(!_superViewBottomConstraint){
            _superViewBottomConstraint = [self.view.bottomAnchor constraintEqualToAnchor:self.view.superview.safeAreaLayoutGuide.bottomAnchor constant:0];
            _superViewBottomConstraint.active = YES;
        }
    }
    else{
        _collectionViewHeightConstraint.constant = contentHeight;
    }
    
    if([self numberOfRowsInCollectionView] == 1) layout.sectionInset = UIEdgeInsetsMake(50, _horizontalPadding, 0, _horizontalPadding);
    else if([self numberOfRowsInCollectionView] == 2) layout.sectionInset = UIEdgeInsetsMake(17, _horizontalPadding, 0, _horizontalPadding);
    else layout.sectionInset = UIEdgeInsetsMake(10, _horizontalPadding, 0, _horizontalPadding);
    //if(contentExceedsView) layout.sectionInset = UIEdgeInsetsMake(7, _horizontalPadding, 0, _horizontalPadding);
}

#pragma mark - UICollectionView DataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.items.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {

    HostCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"HostCell" forIndexPath:indexPath];
    TemporaryHost *host = self.items[indexPath.item];
    [cell configureWithHost:host];
    return cell;
}

#pragma mark - UICollectionViewDelegateFlowLayout

- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return self.cellSize;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView
                   layout:(UICollectionViewLayout *)collectionViewLayout
minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return self.interItemMinimumSpacing;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView
                   layout:(UICollectionViewLayout *)collectionViewLayout
minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return self.minimumLineSpacing;
}

@end

File: VoidLink/ViewControllers/HostCollectionViewController.swift
//
//  HostCollectionViewController.swift
//  VoidLink
//
//  Created by True砖家 on 2025/5/28.
//  Copyright 2025 True砖家 @ Bilibili. All rights reserved.
//

import UIKit

@objcMembers
class HostCell: UICollectionViewCell {
    var cardView: HostCardView?
    private weak var parentVC: UIViewController?

    override func prepareForReuse() {
        super.prepareForReuse()
        cardView?.removeFromSuperview()
        cardView = nil
    }

    private func getHostCardSizeFactor() -> CGFloat {
        let dummyHost = TemporaryHost()
        let dummyCard = HostCardView(host: dummyHost)
        return contentView.bounds.size.height / dummyCard.size.height
    }

    private func viewController() -> UIViewController? {
        var responder: UIResponder? = self
        while let current = responder {
            if let viewController = current as? UIViewController {
                return viewController
            }
            responder = current.next
        }
        return nil
    }

    private func assignDelegateForHostCard() {
        parentVC = viewController()?.parent
        if let parentVC, parentVC.conforms(to: HostCardActionDelegate.self) {
            cardView?.delegate = parentVC as? HostCardActionDelegate
        }
    }

    override func didMoveToWindow() {
        super.didMoveToWindow()
        assignDelegateForHostCard()
    }

    @objc(configureWithHost:)
    func configure(with host: TemporaryHost) {
        if cardView == nil {
            cardView = HostCardView(host: host, andSizeFactor: getHostCardSizeFactor())
            assignDelegateForHostCard()
            if let cardView, cardView.superview == nil {
                contentView.addSubview(cardView)
                NSLayoutConstraint.activate([
                    cardView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
                    cardView.topAnchor.constraint(equalTo: contentView.topAnchor),
                ])
            }
        }
    }
}

@objcMembers
class HostCollectionViewController: UICollectionViewController, UICollectionViewDelegateFlowLayout {
    var interItemMinimumSpacing: CGFloat = 10
    var minimumLineSpacing: CGFloat = 10
    var cellSize: CGSize = .zero
    private(set) var items = NSMutableArray()

    private var collectionViewHeightConstraint: NSLayoutConstraint?
    private var superViewBottomConstraint: NSLayoutConstraint?
    private let flowLayout: UICollectionViewFlowLayout
    private var horizontalPadding: CGFloat

    @objc init() {
        let layout = UICollectionViewFlowLayout()
        flowLayout = layout

        switch UIDevice.current.userInterfaceIdiom {
        case .phone:
            horizontalPadding = 5
        case .pad:
            fallthrough
        default:
            horizontalPadding = 75
        }

        layout.sectionInset = UIEdgeInsets(top: 7, left: horizontalPadding, bottom: 0, right: horizontalPadding)
        super.init(collectionViewLayout: layout)

        collectionViewHeightConstraint = collectionView.heightAnchor.constraint(equalToConstant: 50)
        collectionViewHeightConstraint?.isActive = true
    }

    required init?(coder: NSCoder) {
        flowLayout = UICollectionViewFlowLayout()
        switch UIDevice.current.userInterfaceIdiom {
        case .phone:
            horizontalPadding = 5
        case .pad:
            fallthrough
        default:
            horizontalPadding = 75
        }
        super.init(coder: coder)
    }

    func updateTheme() {
        collectionView.backgroundColor = ThemeManager.hostViewBackgroundColor
        for case let cell as HostCell in collectionView.visibleCells {
            cell.cardView?.updateTheme()
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        collectionView.register(HostCell.self, forCellWithReuseIdentifier: "HostCell")
        collectionView.alwaysBounceVertical = false
        collectionView.showsVerticalScrollIndicator = false
        updateTheme()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }

    override func didMove(toParent parent: UIViewController?) {
        super.didMove(toParent: parent)
    }

    func addHost(_ host: TemporaryHost) {
        if !items.contains(host) {
            items.add(host)
            collectionView.reloadData()
        }
    }

    func removeHost(_ host: TemporaryHost) {
        if items.contains(host) {
            items.remove(host)
            collectionView.reloadData()
        }
    }

    func removeLastItem() {
        if items.count > 0 {
            items.removeLastObject()
            collectionView.reloadData()
        }
    }

    private func numberOfRowsInCollectionView() -> Int {
        let itemCount = collectionView.numberOfItems(inSection: 0)
        if itemCount == 0 {
            return 0
        }

        var rowYs = Set<NSNumber>()
        for index in 0..<itemCount {
            let indexPath = IndexPath(item: index, section: 0)
            if let attr = flowLayout.layoutAttributesForItem(at: indexPath) {
                let y = attr.frame.minY
                rowYs.insert(NSNumber(value: Double(round(y))))
            }
        }

        return rowYs.count
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

        let contentHeight = collectionView.collectionViewLayout.collectionViewContentSize.height
        if let superview = view.superview {
            let contentExceedsView = contentHeight > superview.bounds.size.height - view.frame.origin.y
            if contentExceedsView {
                if superViewBottomConstraint == nil {
                    superViewBottomConstraint = view.bottomAnchor.constraint(equalTo: superview.safeAreaLayoutGuide.bottomAnchor)
                    superViewBottomConstraint?.isActive = true
                }
            } else {
                collectionViewHeightConstraint?.constant = contentHeight
            }
        } else {
            collectionViewHeightConstraint?.constant = contentHeight
        }

        switch numberOfRowsInCollectionView() {
        case 1:
            flowLayout.sectionInset = UIEdgeInsets(top: 50, left: horizontalPadding, bottom: 0, right: horizontalPadding)
        case 2:
            flowLayout.sectionInset = UIEdgeInsets(top: 17, left: horizontalPadding, bottom: 0, right: horizontalPadding)
        default:
            flowLayout.sectionInset = UIEdgeInsets(top: 10, left: horizontalPadding, bottom: 0, right: horizontalPadding)
        }
    }

    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        items.count
    }

    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "HostCell", for: indexPath) as! HostCell
        if let host = items[indexPath.item] as? TemporaryHost {
            cell.configure(with: host)
        }
        if #available(iOS 13.0, *) {
            applyControllerNavigationHighlight(to: cell, highlighted: indexPath == controllerNavigationSelectedIndexPath)
        }
        return cell
    }

    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        sizeForItemAt indexPath: IndexPath) -> CGSize {
        cellSize
    }

    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        interItemMinimumSpacing
    }

    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        minimumLineSpacing
    }
}
